jobs/build/appregistry/Jenkinsfile

```diff
@@ -113,37 +119,125 @@ node {
     def workDir = "${env.WORKSPACE}/workDir"
     buildlib.cleanWorkdir(workDir)
 
-    currentBuild.description = "Collecting appregistry images for ${params.BUILD_VERSION}"
-    currentBuild.displayName += " - ${params.BUILD_VERSION}"
+    def buildVersion = params.BUILD_VERSION
+
+    currentBuild.description = "Collecting appregistry images for ${buildVersion}"
+    currentBuild.displayName += " - ${buildVersion}"
 
     def skipPush = params.SKIP_PUSH
 
+    def validate = { params ->
+        if (params.STREAM in ["stage", "prod"]) {
+            if (!params.SOURCE_ADVISORIES || !params.TARGET_ADVISORY) {
+                currentBuild.description += """\n
+                ERROR: SOURCE_ADVISORIES and TARGET_ADVISORY parameters are required for selected STREAM.
+                """
+                return false
+            }
+        }
+        return true
+    }
+
+    def doozer = { cmd ->
+        buildlib.doozer("--working-dir ${workDir} -g openshift-${buildVersion} ${cmd}", [capture: true])
+    }
+
+    def elliott = { cmd ->
+        buildlib.elliott("-g openshift-${buildVersion} ${cmd}", [capture: true])
+    }
+
+    def getImagesData = { include ->
+        if (include) {
+            include = "--images " + commonlib.cleanCommaList(include)
+        }
+        doozer """
+            ${include}
+            images:print
+            --label 'com.redhat.delivery.appregistry'
+            --short '{label},{name},{build},{version},{component}'
+        """
+    }
+
+    def fetchNVRsFromAdvisories = { advisories ->
+        commonlib.cleanCommaList(advisories).split(",").collect { advisory ->
+            readJSON(text: elliott("get --json - ${advisory}")).errata_builds.values().flatten()
+        }.flatten()
+    }
+
+    def replaceNVRsByFoundInAdvisories = { images, advisoriesNVRs ->
```

Comment:
This part is like a hack and I think it is unnecessary and error-prone. I am thinking why you need this function for replacing rather than directly use the return value `advisoriesNVRs` of `fetchNVRsFromAdvisories()`. Here is my thought:

IIUC, it is used to
1. filter out operators that are not in the list of `IMAGES` from `advisoriesNVRs`, where `IMAGES` are dist-git names (rather than brew component name).
2. supply arguments for `elliott operator-metadata:latest-build NAME`, where `NAME` is dist-git name.

So the root cause is you want to use this function to get dist-git name mapping for operators on advisory, right?

For item 1, considering why this `IMAGES` parameter exists: To build metadata for a subset of operators in case of previous failed builds. In my opinion, it is better to change the parameter to something like `OPERATOR_NVRS` to let user explicitly specify operator NVRs rather than dist-git names. Building with dist-git names sounds like the behavior is up to the time you run the job.

For item 2, I think `doozer operator-metadata:build` should be updated to give you a NVR to tracker later, so that you can directly use the returned NVR rather than fetch with `doozer operator-metadata:latest-build` afterwards because the latter also depends on timing and is possible to attach the wrong build.

Hence, I'd like to propose a change both to the PR and the doozer command for a more robust enhancement.
 
@sosiouxme Would also hear the thoughts from you.

Reply:
Let me take this piece by piece as it's definitely complicated... and complicated logic isn't a good sign, but the domain itself really is a bit convoluted. Probably the dual branches of logic for dev vs stage/prod are the most confusing, and this may be a landmine for anyone that tries to edit this...

> I am thinking why you need this function for replacing rather than directly use the return value advisoriesNVRs of fetchNVRsFromAdvisories(). 

The first problem to solve is determining which images are even OLM operators (only a few extras are). The doozer command does that for all or a subset of configured images (identified by distgit which is our primary key). It looks only at the config and Dockerfile, not what has actually been built.

That's a bit dev-focused, but I don't want to completely invert the logic for non-dev and work from NVR back to distgit, as a separate branch of logic. I think actually the only thing I'd change here (aside from commenting these "methods") is the name. It's really just a lookup using the components doozer identified as operators. How about:
```suggestion
    def findImageNVRsInAdvisories = { images, advisoriesNVRs ->
```
Probably best to also rename `processImages` to something more specific like `parseAndFilterOperators` as it's not just processing the output but also filtering it to just operators.

> So the root cause is you want to use this function to get dist-git name mapping for operators on advisory, right?

Well, and to determine which things are operators, whether we have advisories to draw from or not.

> For item 1, considering why this IMAGES parameter exists: To build metadata for a subset of operators in case of previous failed builds. 

I'd say rather that it exists to limit the scope of the job to those images, whether it's looking them up, rebuilding metadata, or pushing/adding it.

> In my opinion, it is better to change the parameter to something like OPERATOR_NVRS to let user explicitly specify operator NVRs rather than dist-git names. Building with dist-git names sounds like the behavior is up to the time you run the job.

That would force the user to look up NVRs for stuff. Humans are bad at that, tools should do that. Humans are thinking in terms of abstract images, not specific builds. The distgit is our primary key.

> For item 2, I think doozer operator-metadata:build should be updated to give you a NVR back for tracking later, so that you can directly use the returned NVR rather than fetch with doozer operator-metadata:latest-build afterwards because the latter also depends on timing and is possible to attach the wrong build.

I agree I'd like to see that enhancement, although doozer doesn't know beforehand the NVR it is building (the release is not specified), and determining it from the successful task is awkward (maybe you can parse `koji_builds = ...` out of the result, `brew taskinfo` gives you nothing).

But I think that's a nice to have, not a need, because we're going to be looking up the metadata container that matches a particular operator NVR and stream, and we should never have two of those going at the same time, so there shouldn't be a timing issue.

Reply:
@sosiouxme Thanks for the detailed explanation. Those are only some thoughts when I review the PR, but don't donate a `-1`. Some improvements can definitely be done piece by piece.